constant interpreter: enough for basic integer operations

marcrasi opened the pull request:
Implements a constant interpreter that can deal with basic integer operations.

Summary of the features that it includes:
* builtin integer values, and builtin integer insts
* struct and tuple values, and insts that construct and extract them (necessary to use stdlib integers)
* function referencing and application (necessary to call stdlib integer functions)
* error handling data structures and logic, for telling you why your value is not evaluatable
* metatype values (not necessary for integers, but it's only a few extra lines, so I thought it would be more trouble than it's worth to put them in a separate PR)
* conditional branches (ditto)

This is a part of the https://github.com/apple/swift/pull/19579 mega-patch. See there if you're interested in what comes next.

Commit:
constant interpreter: basic integer operations

Implements a constant interpreter that can deal with basic integer operations.

Summary of the features that it includes:
* builtin integer values, and builtin integer insts
* struct and tuple values, and insts that construct and extract them (necessary to use stdlib integers)
* function referencing and application (necessary to call stdlib integer functions)
* error handling data structures and logic, for telling you why your value is not evaluatable
* metatype values (not necessary for integers, but it's only a few extra lines, so I thought it would be more trouble than it's worth to put them in a separate PR)
* conditional branches (ditto)

ravikandhadai commented on include/swift/SILOptimizer/Utils/ConstExpr.h:
```diff
@@ -0,0 +1,86 @@
+// values and return constant values.  These constants may be integer, floating
+// point, and string values, or arrays thereof (up to 1024 elements).  We allow
+// abstractions to be built out of fragile structs and tuples.
+//
```
@marcrasi  

> These constants may be integer, floating point, and string values, or arrays thereof (up to 1024 elements).

Should we remove this comment Line as we don't have array/string in this commit? Is the limit of 1024 elements imposed somewhere in the array extensions? If so, should we really commit to this constant?

ravikandhadai commented on lib/SIL/SILConstants.cpp:
```diff
@@ -0,0 +1,395 @@
+  case RK_Unknown:
+  case RK_Metatype:
+  case RK_Function:
+  case RK_IntegerInline:
```
@marcrasi  Seems like this function will have an assert failure inside `getIntegerValue()` if we call it with any of the first four representation kinds.  I suppose there is no reason why one should clone the first four cases as their values will probably not change at all.  In that case, can we have a default case that fails with an assert and a reasonable error message (like "cloning of this representation kind ... is not supported"? Then we can just remove the first four cases.

ravikandhadai commented on lib/SIL/SILConstants.cpp:
```diff
@@ -0,0 +1,395 @@
+/// This assumes that the elements lifetime has been managed for this.
+SymbolicValue SymbolicValue::getAggregate(ArrayRef<SymbolicValue> elements,
+                                          llvm::BumpPtrAllocator &allocator) {
+  // Copy the integers from the APInt into the bump pointer.
```
@marcrasi The comments are referring to ints. Probably this should be updated for arrays.

ravikandhadai commented on lib/SIL/SILConstants.cpp:
```diff
@@ -0,0 +1,395 @@
+    return loc;
+
+  // Zip through inlined call site information that came from the
+  // implementation guts of the tensor library.  We want to report the
```
@marcrasi Can we rename tensor library to just library?

ravikandhadai commented on lib/SIL/SILConstants.cpp:
```diff
@@ -0,0 +1,395 @@
+    break;
+  case UnknownReason::TooManyInstructions:
+    // TODO: Should pop up a level of the stack trace.
+    error = "expression is too large to evaluate at compile-time";
```
@marcrasi  Can we reword this error? Here large seems to refer to the syntactic size of the expression in the source file. It could be a small expression but may still exceed the instruction limit e.g. if it is recursive, right? Should we reword this to something more explicit like "Exceeded instruction limit: 512  when evaluating the expression at compile time"?

ravikandhadai commented on lib/SIL/SILConstants.cpp:
```diff
@@ -0,0 +1,395 @@
+      SM.getLineNumber(fallbackLoc.getSourceLoc());
+  for (auto &sourceLoc : llvm::reverse(getUnknownCallStack())) {
+    // Skip known sources.
+    if (!sourceLoc.isValid())
```
@marcrasi  Skip "unknown" sources?

ravikandhadai commented on lib/SILOptimizer/Mandatory/DataflowDiagnostics.cpp:
```diff
@@ -125,10 +165,12 @@ class EmitDFDiagnostics : public SILFunctionTransform {
       return;
 
     SILModule &M = getFunction()->getModule();
+    ConstExprEvaluator constantEvaluator(M);
```
@marcrasi  I wonder whether it would be better to create `ConstExprEvaluator` and invoke `diagnosePoundAssert` only when the experimental -enable-static-assert option is enabled.

marcrasi replied:
yes, that sounds good. will do.

ravikandhadai commented on lib/SILOptimizer/Utils/ConstExpr.cpp:
```diff
@@ -0,0 +1,730 @@
+/// callee in a call chain to represent the constant values given the set of
+/// formal parameters that callee was invoked with.
+class ConstExprFunctionState {
+  /// This is the evaluator we put bump pointer allocated values into.
```
@marcrasi Can we make this comment a bit more clear? The evaluator holds a reference to a bump allocator and also provides utility functions like getUnknown() etc.

ravikandhadai commented on lib/SILOptimizer/Utils/ConstExpr.cpp:
```diff
@@ -0,0 +1,730 @@
+    if (!operand.isConstant())
+      return operand;
+
+    // TODO: SUCheckedConversion/USCheckedConversion
```
@marcrasi  Is this SUCheckerConversion is what is being implemented in the following lines? Is it a TODO still?

marcrasi replied:
It's still a TODO. The following lines implement "xxxCheckedTrunc", not "xxxCheckedConversion".

"xxxCheckedConversion" seems to be a builtin that exists but that the stdlib never calls.

ravikandhadai commented on lib/SILOptimizer/Utils/ConstExpr.cpp:
```diff
@@ -0,0 +1,730 @@
+  return evaluator.getUnknown((SILInstruction *)apply, UnknownReason::Default);
+}
+
+// TODO: Refactor this to someplace common, this is defined in Devirtualize.cpp.
```
@marcrasi  This comment is a left-over from other patches right? It should not be removed.

marcrasi replied:
yeah, it shouldn't be in this patch. will remove.

ravikandhadai commented on lib/SILOptimizer/Utils/ConstExpr.cpp:
```diff
@@ -0,0 +1,730 @@
+  SubstitutionMap calleeSubMap;
+
+  // Now that have successfully folded all of the parameters, we can evaluate
+  // the call.
```
@marcrasi  Now the "we" have

ravikandhadai commented on lib/SILOptimizer/Utils/ConstExpr.cpp:
```diff
@@ -0,0 +1,730 @@
+  if (isa<DebugValueInst>(inst) || isa<DebugValueAddrInst>(inst) ||
+      isa<EndAccessInst>(inst) ||
+      // Constant have no important state.
+      isa<DestroyAddrInst>(inst) || isa<RetainValueInst>(inst) ||
```
@marcrasi  Can we just say in comments that "the interpreter doesn't model these memory management instructions and hence skips them.", instead of "Constant have no important state"

ravikandhadai commented on lib/SILOptimizer/Utils/ConstExpr.cpp:
```diff
@@ -0,0 +1,730 @@
+        return val;
+
+      // If we got a constant value, then we're good.  Set up the normal result
+      // values as well any indirect results.
```
as well "as"

ravikandhadai commented on lib/SILOptimizer/Utils/ConstExpr.cpp:
```diff
@@ -0,0 +1,730 @@
+    auto symVal = state.getConstantValue(v);
+    results.push_back(symVal);
+
+    // Reset the execution limit back to zero for each subsexpression we look
```
subsexpression -> subexpression

ravikandhadai commented on lib/SILOptimizer/Utils/ConstExpr.cpp:
```diff
@@ -0,0 +1,730 @@
+///
+/// TODO: Return information about which callees were found to be
+/// constexprs, which would allow the caller to delete dead calls to them
+/// that occur after after folding them.
```
after after -> after

ravikandhadai commented on test/SILOptimizer/pound_assert.swift:
```diff
@@ -0,0 +1,146 @@
+  // TODO: I think that this should get rejected. Otherwise, constness of vars
+  // depends on what you do to them, which is confusing. Perhaps some other
+  // const folder is getting to this before ConstExpr.cpp does?
+  var topLevelVar = 1 // expected-warning {{never mutated}}
```
@marcrasi The answer to the question in the comments is Yes. The `DiagnosticConstantPropagation` phase runs before this phase and it folds this constant. This I think should be okay, it only makes #assert's a little bit more powerful ;-)

ravikandhadai commented on include/swift/SIL/SILConstants.h:
```diff
@@ -0,0 +1,246 @@
+    RK_IntegerInline,
+
+    /// This value is an array, struct, or tuple of constants.  This is
+    /// tracked by the "aggregate" member of the value union.
```
@marcrasi  "This value is an array, ..." -> For arrays, we will have a different representation kind, right? Guess array should be removed from this comment.

marcrasi commented:
I'm partway through rebasing this onto latest changes and addressing your comments, so I replied to some and marked some as resolved. I haven't uploaded my changes yet so no need to take a look or respond yet. I'm going to stop now and do a bit more tomorrow.

marcrasi commented on test/SILOptimizer/pound_assert.swift:
```diff
@@ -0,0 +1,146 @@
+func test_trapsAndOverflows() {
+  // The error message below is generated by the traditional constant folder.
+  // expected-error @+1 {{arithmetic operation '124 + 92' (on type 'Int8') results in an overflow}}
+  #assert((124 as Int8) + 92 < 42)
```
It seems that the other constant folder is replacing the condition in this case, but leaving the condition alone in the next case. I have added some comments explaining that this is happening, but I haven't investigated why the other constant folder is behaving like that.

ravikandhadai replied:
@marcrasi  I see. Thanks for the clarification. Probably the diagnostic constant propagation, in this case, replaces the addition with a truncated value (besides recording the diagnostics). So the interpreter never sees the addition. Btw, a minor comment is that I think it would be better in these comments to use the term "interpreter"  instead of constant folder for #assert. It may create less confusion if we use the terminology "folder" for something that actually changes the code, whereas the interpreter doesn't.  We may want to eventually rewrite the constant folder using the interpreter, but still I think the folding will be done by the constant folder phase.

marcrasi replied:
That sounds like a good rewording. I have updated the comment.

marcrasi commented on lib/SILOptimizer/Utils/ConstExpr.cpp:
```diff
@@ -0,0 +1,730 @@
+    // TODO: We can/should diagnose statically detectable integer overflow
+    // errors and subsume the ConstantFolding.cpp mandatory SIL pass.
+    auto IntCheckedTruncFn = [&](bool srcSigned,
+                                 bool dstSigned) -> SymbolicValue {
```
Yes, that sounds like an improvement. fyi, dstSigned already was used on the line that calculates `reextended`.

marcrasi commented:
Okay, I have rebased this and addressed this round of comments!

For the rebase, I had to change SILConstants.h to accept an ASTContext instead of a BumpPtrAllocator because ASTContext's getAllocator() became private.

ravikandhadai commented:
@marcrasi  Thanks for making these changes. It looks good to me.

marcrasi commented:
I rebased this one to master now that the previous part is merged. Is this ready to test and merge?

ravikandhadai commented:
@marcrasi  Thanks Marc. It would be good to have @devincoughlin  review. Let us wait a bit for that. Let me run the tests in the meantime.

ravikandhadai commented:
@swift-ci Please test

ravikandhadai commented:
@swift-ci Please test macOS platform

ravikandhadai commented:
@swift-ci Please benchmark

devincoughlin reviewed (commented):
This looks good to me. Some minor requests inline and a suggestion to add a SIL-level test.

marcrasi commented on lib/SIL/SILConstants.cpp:
```diff
@@ -0,0 +1,405 @@
+    // Also skip notes that point to the same line as the original error, for
+    // example in:
+    //   #assert(foo(bar()))
+    // it is not useful to get three diagnostics referring to the same line.
```
I can't think of a situation where this logic does something useful, so I'll just remove it.

The example in the comments seems to suggest that `#assert(foo(bar()))` will emit two notes when `bar()` is not const-evaluable: one for `bar` and one for `foo`. But that is not actually what happens -- it seems that the unknown that comes out of `bar()` is the one that emits the notes, so only a note for `bar()` gets emitted. And even if it did emit two notes, I agree that that would be reasonable.

marcrasi commented on lib/SIL/SILConstants.cpp:
```diff
@@ -0,0 +1,405 @@
+
+/// Given that this is an 'Unknown' value, emit diagnostic notes providing
+/// context about what the problem is.
+void SymbolicValue::emitUnknownDiagnosticNotes(SILLocation fallbackLoc) {
```
I don't want to move this logic into a pass because there may be multiple different passes that want to use it to emit diagnostics about unknown values. (e.g. some of the tensorflow passes need to const-eval certain things and emit diagnostics when the const-eval fails).

Having it in SymbolicValue seems like the easiest way for passes who use it to call it, and I don't see much harm in leaving it here. Alternatively, I could turn it into a standalone utility function. Maybe `emitUnknownDiagnosticNotes(SILConstant unknown, SILLocation fallbackLoc)` in `SILOptimizer/Utils/ConstantDiagnostics`? What do you prefer?

marcrasi commented on lib/SIL/SILConstants.cpp:
```diff
@@ -0,0 +1,395 @@
+  std::string error;
+  switch (getUnknownReason()) {
+  case UnknownReason::Default:
+    error = "could not fold operation";
```
will do

marcrasi commented on lib/SILOptimizer/Utils/ConstExpr.cpp:
```diff
@@ -0,0 +1,725 @@
+        numInstEvaluated(numInstEvaluated) {}
+
+  void setValue(SILValue value, SymbolicValue symVal) {
+    calculatedValues.insert({value, symVal});
```
will do. actually, `Address` kinds don't exist until the next PR (https://github.com/apple/swift/pull/19657/files) so I'll just add a commented-out assertion with a TODO to uncomment it.

marcrasi commented on include/swift/SIL/SILConstants.h:
```diff
@@ -0,0 +1,250 @@
+  void dump() const;
+};
+
+static_assert(sizeof(SymbolicValue) == 2 * sizeof(void *),
```
will do

marcrasi commented on test/SILOptimizer/pound_assert.swift:
```diff
@@ -0,0 +1,155 @@
+// RUN: %target-swift-frontend -enable-experimental-static-assert -emit-sil %s -verify -Xllvm -debug -Xllvm -debug-only -Xllvm ConstExpr
```
will do

marcrasi commented on lib/SIL/SILConstants.cpp:
```diff
@@ -0,0 +1,405 @@
+
+  assert(representationKind == RK_Integer);
+  auto numBits = aux.integer_bitwidth;
+  auto numWords = (numBits + 63) / 64;
```
will do

marcrasi commented on include/swift/SIL/SILConstants.h:
```diff
@@ -0,0 +1,246 @@
+
+  union {
+    /// This is the reason code for RK_Unknown values.
+    UnknownReason unknown_reason : 32;
```
will do

marcrasi commented on include/swift/SIL/SILConstants.h:
```diff
@@ -0,0 +1,246 @@
+
+    /// This is the number of elements for an RK_Aggregate representation.
+    unsigned aggregate_numElements;
+  } aux;
```
will do

marcrasi commented on lib/SIL/SILConstants.cpp:
```diff
@@ -0,0 +1,395 @@
+  UnknownReason reason;
+
+  /// The number of elements in the call stack.
+  unsigned call_stack_size;
```
will change

Commit:
address comments

ravikandhadai commented:
@swift-ci please smoke test

devincoughlin commented on lib/SILOptimizer/Utils/ConstExpr.cpp:
```diff
@@ -0,0 +1,725 @@
+/// function.  If so, collect its arguments as constants, fold it and return
+/// None.  If not, mark the results as Unknown, and return an Unknown with
+/// information about the error.
+llvm::Optional<SymbolicValue>
```
Yes, that sounds good.

devincoughlin commented:
Looks good to me. Thanks!

ravikandhadai commented:
@swift-ci Please test

ravikandhadai commented:
@swift-ci please test and merge
